enhancements/ingress/client-tls.md

```diff
@@ -0,0 +1,315 @@
+custom CA certificate bundle with a CA certificate using the X509v3 extension to
+specify one or more CRL destination points in the certificate.  The operator
+configures the IngressController to request and verify client certificates using
+the provided CA certificate bundle and reject clients that do not present valid
+client certificates.  The operator periodically updates the CRL using the
+destination points.
+
+### Implementation Details
+
+OpenShift Router configures HAProxy using a configuration template.  The
+template uses environment variables as input parameters.  Most of the basic
+functionality for configuring client TLS already exists in the template and can
+be configured using well defined environment variables.  The client certificate
+policy can be configured using the `ROUTER_MUTUAL_TLS_AUTH` variable.  The path
+to a CA certificate bundle can be specified using the
+`ROUTER_MUTUAL_TLS_AUTH_CA` variable, and similarly, the path to a file
+containing associated CRLs can be specified using the
+`ROUTER_MUTUAL_TLS_AUTH_CRL` variable.  A pattern for filtering client
+certificates by their subjects can be specified using the
+`ROUTER_MUTUAL_TLS_AUTH_FILTER` variable.  The ingress operator uses these
+environment variables to configure client TLS.
+
+To inject the CA certificate bundle into the router pod, the operator adds a
+configmap volume and volume mount and specifies the mount path using
+`ROUTER_MUTUAL_TLS_AUTH_CA` in the router deployment.
+
+To inject the CRL file into the router pod, the operator must first scan the
+certificates in the IngressController's CA certificate bundle configmap to
+determine whether any certificates specify any CRL destination endpoints.  If
+none does, then no CRL file is needed, and so no further configuration is
+required for CRLs.  However, if any certificate does specify any CRL
+distribution points, then the operator configures an additional configmap volume
+and volume mount for the CRLs and specifies the mount path using
+`ROUTER_MUTUAL_TLS_AUTH_CRL`.  The operator configures the deployment so that
+its pods cannot start until the CRL configmap exists.  This is necessary to
+ensure that revoked certificates are not permitted while the CRL is downloaded.
+
+A separate controller in the operator scans client CA certificate bundle
+configmaps, and if it finds that any certificates in a configmap specify any CRL
+destination points, then the controller downloads the CRLs and creates or
+updates the CRL configmap for the associated IngressController.
+
+Finally, if `spec.clientTLS.allowedSubjectPatterns` is nonempty, the operator
+combines the patterns therein and sets the result as the value for
+`ROUTER_MUTUAL_TLS_AUTH_FILTER` in the router deployment.
+
+### Risks and Mitigations
+
+If the underlying IngressController implementation were to change away from
+HAProxy to a different implementation, we would need to ensure that the new
+implementation supported the same capabilities.
```

Comment:
s/supported/supports

Reply:
I'm using "supported" for the subjunctive mood ("If A were to change from B to C, we would need to ensure that D supported E").